File: python/coins.py
coin_values = [200, 100, 50, 20, 10, 5, 2, 1]
number_coins = [5, 10, 10, 20, 30, 40, 50, 50]

def get_available_amount():
    sum = 0
    for idx in range(len(coin_values)):
        sum += coin_values[idx] * number_coins[idx]
    return sum

def get_change(total_amount):
    if get_available_amount() < total_amount:
        return "not enough money"
    coins = []
    for idx in range(len(coin_values)):
        value = coin_values[idx]
        if total_amount >= value:
            q, _ = divmod(total_amount, value)
            n_coins = q if number_coins[idx] >= q else number_coins[idx]
            total_amount -= n_coins * value 
            coins.append((n_coins, value))
            number_coins[idx] -= n_coins
        if total_amount == 0:
            break
    if total_amount > 0:
        return "not enough money"
    return coins

print(get_change(1234))

print(f"available money: {get_available_amount()}")

File: python/coins.cpp
#include <iostream>
#define NUMBER_COINS 8
using namespace std;

int coin_values[NUMBER_COINS] = {200, 100, 50, 20, 10, 5, 2, 1};
int number_coins[NUMBER_COINS] = {5, 10, 10, 20, 30, 40, 50, 50};

int get_avaliable_amount()
{
    auto sum = 0;
    for(auto i = 0; i < NUMBER_COINS; i++)
    {
        sum += coin_values[i] * number_coins[i];
    }
    return sum;
}


void get_change(int total_amount)
{
    // if get_available_amount() < total_amount:
    //     return "not enough money"
    if(get_avaliable_amount() < total_amount)
    {
        cout << "not enough money" << endl;
        return;
    }

    for(auto i = 0; i < NUMBER_COINS; i++)
    {
        auto value = coin_values[i];
        if(total_amount >= value)
        {
            auto q = total_amount / value;
            auto nr_coins = number_coins[i] >= q ? q: number_coins[i];
            total_amount -= nr_coins * value;
            cout << nr_coins << " coin(s) of " << value << " cents" << endl;
            number_coins[i] -= nr_coins;
        }
        if (total_amount == 0)
        {
            break;
        }
    }

    if (total_amount > 0)
    {
        cout << "not enough money" << endl;
    }
}

int main()
{
    get_change(1234);
    cout << get_avaliable_amount() << endl;    
}